windows/hid.c

```diff
@@ -358,28 +381,84 @@ HID_API_EXPORT const char* HID_API_CALL hid_version_str(void)
 	return HID_API_VERSION_STR;
 }
 
+static void hid_internal_hotplug_init()
+{
+	if (!hid_hotplug_context.mutex) {
+		hid_hotplug_context.mutex = CreateMutexW(NULL, FALSE, NULL);
+		if(!hid_hotplug_context.mutex) {
+			register_global_error(L"Failed to create hotplug management mutex");
+		}
+	}
+}
+
 int HID_API_EXPORT hid_init(void)
 {
 	register_global_error(NULL);
+
 #ifndef HIDAPI_USE_DDK
 	if (!hidapi_initialized) {
 		if (lookup_functions() < 0) {
 			register_global_winapi_error(L"resolve DLL functions");
 			return -1;
-		}
+        }
 		hidapi_initialized = TRUE;
 	}
 #endif
+
+	hid_internal_hotplug_init();
+
 	return 0;
 }
 
+static void hid_internal_hotplug_cleanup()
+{
+	/* Unregister a HID device connection notification when removing the last callback */
+	/* This function is always called inside a locked mutex */
+	if (hid_hotplug_context.hotplug_cbs == NULL) {
+		if (hid_hotplug_context.devs) {
+			/* Cleanup connected device list */
+			hid_free_enumeration(hid_hotplug_context.devs);
+			hid_hotplug_context.devs = NULL;
+		}
+
+		if (hid_hotplug_context.notify_handle) {
+			if (CM_Unregister_Notification(hid_hotplug_context.notify_handle) != CR_SUCCESS) {
+				register_global_error(
+					L"CM_Unregister_Notification failed for Hotplug notification");
+				return;
+			}
+		}
+
+		hid_hotplug_context.notify_handle = NULL;
+	}
+}
+
+static void hid_internal_hotplug_exit()
+{
+	WaitForSingleObject(hid_hotplug_context.mutex, INFINITE);
+	hid_hotplug_callback** current = &hid_hotplug_context.hotplug_cbs
+	/* Reove all callbacks from the list */
```

Comment:
Typo

Reply:
Done. Unfortunatelly, this got copy-pasted into other 2 branches as well...